util/dvsim/Scheduler.py

```diff
@@ -233,44 +104,372 @@ def on_sigint(signal_received, frame):
+            dep_status = self.item_to_status[dep]
+            assert dep_status in ['P', 'F', 'K']
+
+            if item.needs_all_dependencies_passing:
+                if dep_status in ['F', 'K']:
+                    return False
+            else:
+                if dep_status in ['P']:
+                    return True
+
+        return item.needs_all_dependencies_passing
+
+    def _poll(self, hms):
+        '''Check for running items that have finished
+
+        Returns True if something changed.
+
+        '''
+
+        changed = False
+
+        for target in self._scheduled:
+            if target in self._completed:
+                continue
+
+            to_pass = []
+            to_fail = []
+            for item in self._running[target]:
+                status = item.launcher.poll()
+                assert status in ['D', 'P', 'F']
+                if status == 'D':
+                    # Still running
+                    continue
+                elif status == 'P':
+                    log.log(VERBOSE, "[%s]: [%s]: [status] [%s: P]", hms,
+                            target, item.full_name)
+                    to_pass.append(item)
+                else:
+                    log.error("[%s]: [%s]: [status] [%s: F]", hms, target,
+                              item.full_name)
+                    to_fail.append(item)
+
+            for item in to_pass:
+                self._passed[target].add(item)
+                self._running[target].remove(item)
+                self.item_to_status[item] = 'P'
+                self._enqueue_successors(item)
+
+            for item in to_fail:
+                self._failed[target].add(item)
+                self._running[target].remove(item)
+                self.item_to_status[item] = 'F'
+                # It may be possible that a failed item's successor may not
+                # need all of its dependents to pass (if it has other dependent
+                # jobs). Hence we enqueue all successors rather than canceling
+                # them right here. We leave it to `_dispatch()` to figure out
+                # whether an enqueued item can be run or not.
+                self._enqueue_successors(item)
+
+            changed = changed or to_pass or to_fail
+
+        return changed
+
+    def _dispatch(self, hms):
+        '''Dispatch some queued items if possible.'''
+
+        # Compute sum of lengths of all lists in the given dict.
+        def __sum(d):
+            return sum([len(d[k]) for k in d])
+
+        num_slots = min(Scheduler.max_parallel - __sum(self._running),
+                        __sum(self._queued))
+        if num_slots <= 0:
+            return
+
+        for target in self._scheduled:
+            if target in self._completed:
+                continue
+
+            num_slots_per_target = min(
+                num_slots, Scheduler.max_parallel - len(self._running[target]),
+                len(self._queued[target]))
```

Comment:
This is pretty hard to read because of the mess made by the auto-formatter. Maybe define a local variable for the second argument? Also, it's probably simpler not to include the length here: you can just check whether `self._queued[target]` is empty in the test of the while loop below. That also gets rid of the `try/except` block.

Finally, can you call this `num_slots_for_target` instead? We'll compute different values for different targets, and "per" makes it sound like this is something that applies across the board.